lxd/auth/authorization_objects.go

```diff
@@ -0,0 +1,275 @@
+		}
+
+		components = append(components, unescape(escapedComponent))
+	}
+
+	return projectName, components
+}
+
+func (o Object) validate() error {
+	objectType := o.Type()
+	v, ok := objectValidators[objectType]
+	if !ok {
+		return fmt.Errorf("Missing validator for object of type %q", objectType)
+	}
+
+	projectName, identifierElements := o.projectAndElements()
+	if v.requireProject && projectName == "" {
+		return fmt.Errorf("Authorization objects of type %q require a project", objectType)
+	}
+
+	if len(identifierElements) != v.nIdentifierElements {
+		return fmt.Errorf("Authorization objects of type %q require %d components to be uniquely identifiable", objectType, v.nIdentifierElements)
+	}
+
+	return nil
+}
+
+// objectValidator contains fields that can be used to determine if a string is a valid Object.
+type objectValidator struct {
+	nIdentifierElements int
+	requireProject      bool
+}
+
+var objectValidators = map[ObjectType]objectValidator{
+	ObjectTypeUser:          {nIdentifierElements: 1, requireProject: false},
+	ObjectTypeServer:        {nIdentifierElements: 1, requireProject: false},
+	ObjectTypeCertificate:   {nIdentifierElements: 1, requireProject: false},
+	ObjectTypeStoragePool:   {nIdentifierElements: 1, requireProject: false},
+	ObjectTypeProject:       {nIdentifierElements: 0, requireProject: true},
+	ObjectTypeImage:         {nIdentifierElements: 1, requireProject: true},
+	ObjectTypeImageAlias:    {nIdentifierElements: 1, requireProject: true},
+	ObjectTypeInstance:      {nIdentifierElements: 1, requireProject: true},
+	ObjectTypeNetwork:       {nIdentifierElements: 1, requireProject: true},
+	ObjectTypeNetworkACL:    {nIdentifierElements: 1, requireProject: true},
+	ObjectTypeNetworkZone:   {nIdentifierElements: 1, requireProject: true},
+	ObjectTypeProfile:       {nIdentifierElements: 1, requireProject: true},
+	ObjectTypeStorageBucket: {nIdentifierElements: 2, requireProject: true},
+	ObjectTypeStorageVolume: {nIdentifierElements: 3, requireProject: true},
+}
+
+// NewObject returns an Object of the given type. The passed in arguments must be in the correct
+// order (as found in the URL for the resource). This function will error if an invalid object type is
+// given, or if the correct number of arguments is not passed in.
+func NewObject(objectType ObjectType, projectName string, identifierElements ...string) (Object, error) {
+	v, ok := objectValidators[objectType]
+	if !ok {
+		return "", fmt.Errorf("Missing validator for object of type %q", objectType)
+	}
+
+	if v.requireProject && projectName == "" {
+		return "", fmt.Errorf("Authorization objects of type %q require a project", objectType)
+	}
+
+	if len(identifierElements) != v.nIdentifierElements {
+		return "", fmt.Errorf("Authorization objects of type %q require %d components to be uniquely identifiable", objectType, v.nIdentifierElements)
+	}
+
+	builder := strings.Builder{}
+	builder.WriteString(string(objectType))
+	builder.WriteString(":")
```

Comment:
Please could you add a comment here explaining the choice of delimiter and how we avoid that character appearing in the identifierElements?

Reply:
We could define `:` and `/` below as constants with explanations as to why they are used.

Reply:
Done.